In [73]:
import pandas as pd
import numpy as np
import matplotlib as plt
import plotly.express as px

What is the difference between data engineer, data scientist, and software engineers in the tech industry?

How is the data science industry changing over time?


#### WHAT ARE THE DEMOGRAPHICS OF THE DATA SCIENCE INDUSTRY AND HOW HAS THAT CHANGED OVER TIME? 2018-(Q1, Q2, Q3, Q4, Q9, Q5)
1.) Gender/Pay breakdown
2.) Education/Age breakdown 
3.) Undergraduate degree
3.) Country distribution over time how is that changing


#### HOW ARE DATA SCIENTISTS BECOMING DATA SCIENTISTS? 2018-(Q6, Q8, Q18, Q24, Q26, Q35, Q36, Q37, Q38, Q39, Q40)
1.) 

#### WHAT TOOLS/PROGRAMMING LANGUAGES/ENVIRONMENTS DO DATA SCIENTISTS USE AND HOW HAS THAT CHANGED OVER TIME? 2018-(Q12, Q13, Q14, Q15, Q16, Q17, Q19, Q20, Q21, Q22, Q49, Q27, Q28, Q29, Q30, Q31, Q32, Q33, Q49, Q50)
6.) Most used Programming Language? What languages are recommended to learn?
7.) Is data science happening more often in the cloud or locally? What type of IDEs?
3.) Tools they're using

#### WHAT TYPE OF DATA ARE DATA SCIENTISTS DEALING WITH? 
1.) Size of data
2.) Frequency of data

#### WHAT DO DATA SCIENTISTS DO ON A DAILY BASIS? The Role: 2018-(Q11, Q47, Q48, Q50, Q51, Q23, Q34, Q47)
1.) % breakdown of modeling, cleaning, wrangling, communicating, etc.
5.) Activities that are part of their job
2.) Major barriers to accomplishing their work


#### BIAS IN DATA: WHAT ARE WE DOING TO ADDRESS IT? 2018-(Q41, Q43, Q45, Q46, Q48)
1.) Do DS/ML believe in data bias in their algorithms?

#### INDUSTRY BREAKDOWN: WHERE ARE THE MOST DATA SCIENTISTS? IS IT INDUSTRY AGNOSTIC? 2018-(Q7, Q10, Q42, Q44)

# 2019, 2018, 2017 Kaggle ML & DS Survey #
### A Comprehensive Survey Analysis of 24,000 Machine Learning & Data Science Practioners####

This year, as last year, we set out to conduct an industry-wide survey that presents a truly comprehensive view of the state of data science and machine learning. The survey was live for one week in October, and after cleaning the data we finished with 23,859 responses, a 49% increase over last year!

There's a lot to explore here. The results include raw numbers about who is working with data, what’s happening with machine learning in different industries, and the best ways for new data scientists to break into the field. We've published the data in as raw a format as possible without compromising anonymization, which makes it an unusual example of a survey dataset.

The challenge objective: tell a data story about a subset of the data science community represented in this survey, through a combination of both narrative text and data exploration. A “story” could be defined any number of ways, and that’s deliberate. The challenge is to deeply explore (through data) the impact, priorities, or concerns of a specific group of data science and machine learning practitioners. That group can be defined in the macro (for example: anyone who does most of their coding in Python) or the micro (for example: female data science students studying machine learning in masters programs). This is an opportunity to be creative and tell the story of a community you identify with or are passionate about!

Submissions will be evaluated on the following:

Composition - Is there a clear narrative thread to the story that’s articulated and supported by data? The subject should be well defined, well researched, and well supported through the use of data and visualizations.
Originality - Does the reader learn something new through this submission? Or is the reader challenged to think about something in a new way? A great entry will be informative, thought provoking, and fresh all at the same time.
Documentation - Are your code, and kernel, and additional data sources well documented so a reader can understand what you did? Are your sources clearly cited? A high quality analysis should be concise and clear at each step so the rationale is easy to follow and the process is reproducible

In [131]:
df_freeForm = pd.read_csv('kaggle-survey-2018/freeFormResponses.csv')
df_multi_choice = pd.read_csv('kaggle-survey-2018/multipleChoiceResponses.csv')
df_survey_schema = pd.read_csv('kaggle-survey-2018/SurveySchema.csv')


#Plot Global Variables
COLUMN_COLORS = ["#2E648B","#59BAD1","#FAA73F","deepskyblue","darkturquoise","paleturquoise","turquoise"]
BKG_COLOR = ['#EDEDED']
OPACITY = 0.75

#Load 2017 Data 
#df_2017 = pd.read_csv('../2017_data/kaggle-survey-2017/multipleChoiceResponses.csv')

#Load 2019 Data
df_2019 = pd.read_csv('../2019_data/kaggle-survey-2019/multiple_choice_responses.csv')



/Users/adrianlievano/anaconda3/lib/python3.6/site-packages/IPython/core/interactiveshell.py:2728: DtypeWarning:

Columns (25) have mixed types. Specify dtype option on import or set low_memory=False.

/Users/adrianlievano/anaconda3/lib/python3.6/site-packages/IPython/core/interactiveshell.py:2728: DtypeWarning:

Columns (0,2,8,10,21,23,24,25,26,27,28,44,56,64,83,85,87,107,109,123,125,150,157,172,174,194,210,218,219,223,246,249,262,264,276,277,278,279,280,281,282,283,284,285,286,287,288,289,290,304,306,325,326,329,341,368,371,384,385,389,390,391,393,394) have mixed types. Specify dtype option on import or set low_memory=False.

/Users/adrianlievano/anaconda3/lib/python3.6/site-packages/IPython/core/interactiveshell.py:2728: DtypeWarning:

Columns (0,3,7,19,34,47,49,50,51,52,53,54,68,81,94,96,109,115,130,139,147,154,167,180,193,206,219,232,245) have mixed types. Specify dtype option on import or set low_memory=False.



In [132]:
df_2019.head(2)

Time from Start to Finish (seconds)                           Q1  \
0               Duration (in seconds)  What is your age (# years)?   
1                                 510                        22-24   

                                       Q2  \
0  What is your gender? - Selected Choice   
1                                    Male   

                                       Q2_OTHER_TEXT  \
0  What is your gender? - Prefer to self-describe...   
1                                                 -1   

                                          Q3  \
0  In which country do you currently reside?   
1                                     France   

                                                  Q4  \
0  What is the highest level of formal education ...   
1                                    Master’s degree   

                                                  Q5  \
0  Select the title most similar to your current ...   
1                                  Software Engineer   

                                       Q5_OTHER_TEXT  \
0  Select the title most similar to your current ...   
1                                                 -1   

                                                  Q6  \
0  What is the size of the company where you are ...   
1                               1000-9,999 employees   

                                                  Q7  \
0  Approximately how many individuals are respons...   
1                                                  0   

                         ...                          \
0                        ...                           
1                        ...                           

                                          Q34_Part_4  \
0  Which of the following relational database pro...   
1                                                NaN   

                                          Q34_Part_5  \
0  Which of the following relational database pro...   
1                                                NaN   

                                          Q34_Part_6  \
0  Which of the following relational database pro...   
1                                                NaN   

                                          Q34_Part_7  \
0  Which of the following relational database pro...   
1                                                NaN   

                                          Q34_Part_8  \
0  Which of the following relational database pro...   
1                                                NaN   

                                          Q34_Part_9  \
0  Which of the following relational database pro...   
1                                                NaN   

                                         Q34_Part_10  \
0  Which of the following relational database pro...   
1                                                NaN   

                                         Q34_Part_11  \
0  Which of the following relational database pro...   
1                                                NaN   

                                         Q34_Part_12  \
0  Which of the following relational database pro...   
1                                                NaN   

                                      Q34_OTHER_TEXT  
0  Which of the following relational database pro...  
1                                                 -1  

[2 rows x 246 columns]

In [75]:
df_survey_schema.head()

2018 Kaggle Machine Learning and Data Science Survey  \
0                                          Question:     
1                                  # of Respondents:     
2  Who was excluded? (0 = not excluded; 1 = exclu...     
3  If What is your age (# years)? 0-17 Is Selecte...     
4  If What is the highest level of formal educati...     

                                       Q1  \
0  What is your gender? - Selected Choice   
1                                   23860   
2                                       0   
3                                       0   
4                                       0   

                                                 Q10  \
0  Does your current employer incorporate machine...   
1                                              20670   
2                                                  0   
3                                                  1   
4                                                  0   

                                                 Q11  \
0  Select any activities that make up an importan...   
1                                              19518   
2                                                  0   
3                                                  1   
4                                                  0   

                                                 Q12  \
0  What is the primary tool that you use at work ...   
1                                              19199   
2                                                  0   
3                                                  1   
4                                                  0   

                                                 Q13  \
0  Which of the following integrated development ...   
1                                              19117   
2                                                  0   
3                                                  1   
4                                                  0   

                                                 Q14  \
0  Which of the following hosted notebooks have y...   
1                                              18971   
2                                                  0   
3                                                  1   
4                                                  0   

                                                 Q15  \
0  Which of the following cloud computing service...   
1                                              18864   
2                                                  0   
3                                                  1   
4                                                  0   

                                                 Q16  \
0  What programming languages do you use on a reg...   
1                                              18828   
2                                                  0   
3                                                  1   
4                                                  0   

                                                 Q17  \
0  What specific programming language do you use ...   
1                                              15223   
2                                                  0   
3                                                  1   
4                                                  0   

                  ...                  \
0                 ...                   
1                 ...                   
2                 ...                   
3                 ...                   
4                 ...                   

                                                 Q47  \
0  What methods do you prefer for explaining and/...   
1                                              13418   
2                                                  0   
3                                                  1   
4                                                  0   

                                                 Q48  \
0  Do you consider ML models to be "black boxe

In [76]:
columns = df_survey_schema.columns[0:]
len(columns)

52

## Print and format all the questions in the 2018 Kaggle Survey 

print('There are {} questions in this survey'.format(len(df_survey_schema.iloc[0])))

for i in range(len(df_survey_schema.iloc[0])):
    
    print(' ')
    print(columns[i])
    print(df_survey_schema.iloc[0][i])
    print(' ')
    

In [133]:
df_2019 = df_2019[1:]


In [77]:
#Remove row 1 and reset index to 0 to make analysis easier
df_multi_choice = df_multi_choice[1:]
df_multi_choice.head()

Time from Start to Finish (seconds)      Q1 Q1_OTHER_TEXT     Q2  \
1                                 710  Female            -1  45-49   
2                                 434    Male            -1  30-34   
3                                 718  Female            -1  30-34   
4                                 621    Male            -1  35-39   
5                                 731    Male            -1  22-24   

                         Q3                 Q4  \
1  United States of America    Doctoral degree   
2                 Indonesia  Bachelor’s degree   
3  United States of America    Master’s degree   
4  United States of America    Master’s degree   
5                     India    Master’s degree   

                                                  Q5              Q6  \
1                                              Other      Consultant   
2                 Engineering (non-computer focused)           Other   
3      Computer science (software engineering, etc.)  Data Scientist   
4  Social sciences (anthropology, psychology, soc...    Not employed   
5                          Mathematics or statistics    Data Analyst   

  Q6_OTHER_TEXT                         Q7      ...       Q49_OTHER_TEXT  \
1            -1                      Other      ...                   -1   
2             0  Manufacturing/Fabrication      ...                   -1   
3            -1             I am a student      ...                   -1   
4            -1                        NaN      ...                   -1   
5            -1             I am a student      ...                   -1   

  Q50_Part_1          Q50_Part_2                             Q50_Part_3  \
1        NaN                 NaN                                    NaN   
2        NaN                 NaN                                    NaN   
3        NaN  Too time-consuming                                    NaN   
4        NaN                 NaN  Requires too much technical knowledge   
5        NaN  Too time-consuming                                    NaN   

  Q50_Part_4                              Q50_Part_5 Q50_Part_6 Q50_Part_7  \
1        NaN                                     NaN        NaN        NaN   
2        NaN                                     NaN        NaN        NaN   
3        NaN                                     NaN        NaN        NaN   
4        NaN  Not enough incentives to share my work        NaN        NaN   
5        NaN  Not enough incentives to share my work        NaN        NaN   

  Q50_Part_8 Q50_OTHER_TEXT  
1        NaN             -1  
2        NaN             -1  
3        NaN             -1  
4        NaN             -1  
5        NaN             -1  

[5 rows x 395 columns]

for col in df_multi_choice.columns:
    print(col)

In [78]:
df_test = pd.DataFrame({'count' : df_multi_choice.groupby([ "Q1", "Q9"] ).size()}).reset_index().sort_values(by = 'count', ascending = False)
df_test.head()



Q1                                                 Q9  count
36  Male  I do not wish to disclose my approximate yearl...   3759
18  Male                                           0-10,000   3610
19  Male                                          10-20,000   1616
23  Male                                          20-30,000   1181
26  Male                                          30-40,000    933

In [79]:
fig = px.bar(df_test, x = 'Q9', y = 'count', color = 'Q1', barmode = 'group', 
             title= 'Compensation breakdown per Gender Category',
            labels={'Q9':'', 'count':'Total number of responses'},
            opacity = OPACITY)
fig.show()

In [134]:
df_test = pd.DataFrame({'count' : df_2019.groupby([ "Q4"] ).size()}).reset_index().sort_values(by = 'count', ascending = False)
df_test.head()

fig = px.bar(df_test, x = 'Q4', y = 'count', title= 'Education of Respondents in 2019',
            labels={'Q4':'', 'count':'Total number of responses'},
            opacity = OPACITY)
fig.show()

In [80]:
df_test = pd.DataFrame({'count' : df_multi_choice.groupby([ "Q4"] ).size()}).reset_index().sort_values(by = 'count', ascending = False)
df_test.head()

fig = px.bar(df_test, x = 'Q4', y = 'count', title= 'Education of Respondents in 2018',
            labels={'Q4':'', 'count':'Total number of responses'},
            opacity = OPACITY)
fig.show()

## Age Breakdown

In [137]:
df_test = pd.DataFrame({'count' : df_2019.groupby([ "Q5", "Q1"] ).size()}).reset_index().sort_values(by = 'count', ascending = False)
df_test.head()

fig = px.bar(df_test, x = 'Q5', y = 'count', title= 'Age and Role of Survey Respondents in 2019', 
             color = 'Q1' , barmode='group' ,
            labels={'Q5':'', 'count':'Total number of responses'},
            opacity = OPACITY)
fig.show()

In [81]:
df_test = pd.DataFrame({'count' : df_multi_choice.groupby([ "Q6", "Q2"] ).size()}).reset_index().sort_values(by = 'count', ascending = False)
df_test.head()

fig = px.bar(df_test, x = 'Q6', y = 'count', title= 'Age and Role of Survey Respondents in 2018', 
             color = 'Q2' , barmode='group' ,
            labels={'Q6':'', 'count':'Total number of responses'},
            opacity = OPACITY)
fig.show()

Notes:
    1. The figure above is interactive. Drag and drop a box around the role you want to learn more about. The purpose of this figure is to enable readers to explore - at a high level - the data.

In [82]:
df_test = pd.DataFrame({'count' : df_multi_choice.groupby([ "Q3", "Q1"] ).size()}).reset_index().sort_values(by = 'count', ascending = False)
df_test.head()


Q3      Q1  count
67                      India    Male   3719
177  United States of America    Male   3530
28                      China    Male   1337
176  United States of America  Female   1082
113                     Other    Male    849

In [83]:
fig = px.bar(df_test, x = 'Q3', y = 'count', color = 'Q1', barmode = 'group', 
             title= 'Where do Survey Respondents Reside?',
             labels={'Q3':''},
            opacity = OPACITY)
fig.show()

Notes:
    1. The figure above is interactive. Drag and drop a box around the country you want to learn more about. The purpose of this figure is to enable readers to explore - at a high level - the data.

In [140]:
df_test = pd.DataFrame({'count' : df_multi_choice.groupby([ "Q3"] ).size()}).reset_index().sort_values(by = 'count', ascending = False)
df_test.head()


Q3  count
57  United States of America   4716
22                     India   4417
9                      China   1644
37                     Other   1036
45                    Russia    879

In [146]:
fig = px.bar(df_test, x = 'Q3', y = 'count', title= 'Where do Survey Respondents reside?',
            labels={'Q3':'', 'count':'Total number of responses'}, opacity = OPACITY)
fig.show()

Notes:
    1. The figure above is interactive. Drag and drop a box around the country you want to learn more about. The purpose of this figure is to enable readers to explore - at a high level - the data.

## 2018, 2019 Survey Respondents

In [173]:
import plotly.graph_objects as go
from plotly.subplots import make_subplots

top_n= 8
fig = make_subplots(rows=1, cols=2, specs=[[{'type':'domain'}, {'type':'domain'}]])

df_test = pd.DataFrame({'count' : df_multi_choice.groupby([ "Q3"] ).size()}).reset_index().sort_values(by = 'count', ascending = False)

labels = df_test['Q3'].unique()[:top_n]
values = df_test['count'][:top_n]

#use to create 2019 and 2017 pie chart

df_test_2019 = pd.DataFrame({'count' : df_2019.groupby([ "Q3"] ).size()}).reset_index().sort_values(by = 'count', ascending = False)
labels_2019 = df_test_2019['Q3'].unique()[:top_n]
values_2019 = df_test_2019['count'][:top_n]

fig.add_trace(go.Pie(labels = labels, values = values, title_text ='2018'), 1, 1)
fig.add_trace(go.Pie(labels=labels_2019, values=values_2019, title_text ='2019'), 1, 2)

#Use `hole` to create a donut-like pie chart
fig.update_traces(hole=.4, hoverinfo="label+percent+name")
fig.update_layout(title_text="Survey Respondents by Country 2018-2019")
fig.show()



### Key takeaways from Aggregate Data:

    1. The variety of roles, countries, ages, and shows that the Kaggle Data Science community is diverse and international. The 2018 survey shows that the United States only accounts for 29.7% of survey respondents. The two largest countries in Asia (India & China) account for nearly 37.9% of all survey respondents. 
    2. 

Note: For the purposes of this analysis, we will focus on self-described students, data scientists, software engineers, and data engineers in an international context 

In [87]:
df_multi_tech = df_multi_choice[(df_multi_choice['Q6'] == 'Data Scientist') |
                (df_multi_choice['Q6'] == 'Data Engineer') |
                (df_multi_choice['Q6'] == 'Software Engineer')]

df_multi_tech.head(1)

Time from Start to Finish (seconds)      Q1 Q1_OTHER_TEXT     Q2  \
3                                 718  Female            -1  30-34   

                         Q3               Q4  \
3  United States of America  Master’s degree   

                                              Q5              Q6  \
3  Computer science (software engineering, etc.)  Data Scientist   

  Q6_OTHER_TEXT              Q7      ...       Q49_OTHER_TEXT Q50_Part_1  \
3            -1  I am a student      ...                   -1        NaN   

           Q50_Part_2 Q50_Part_3 Q50_Part_4 Q50_Part_5 Q50_Part_6 Q50_Part_7  \
3  Too time-consuming        NaN        NaN        NaN        NaN        NaN   

  Q50_Part_8 Q50_OTHER_TEXT  
3        NaN             -1  

[1 rows x 395 columns]

In [88]:
percent_tech_respondents = (df_multi_tech.shape[0]/df_multi_choice.shape[0])*100 

percent_tech_respondents

33.547089148748896

Note: Data scientists, data engineers, and software engineers account for 33.54% of the total number survey respondents. 

## What is the education breakdown for our group of select respondents? ##

In [89]:
df_test = df_multi_tech.groupby(['Q6','Q4']).size().reset_index(name = 'Percentage')
a = df_test.groupby('Q6')['Percentage'].transform('sum')
df_test['Percentage'] = df_test['Percentage'].div(a).multiply(100)
df_test = df_test.sort_values(by = ['Q6','Percentage'], ascending = False)
df_test.head()

Q6                                                 Q4  \
17  Software Engineer                                    Master’s degree   
14  Software Engineer                                  Bachelor’s degree   
15  Software Engineer                                    Doctoral degree   
20  Software Engineer  Some college/university study without earning ...   
19  Software Engineer                                Professional degree   

    Percentage  
17   43.769968  
14   39.488818  
15    5.910543  
20    4.376997  
19    2.747604

In [90]:
fig = px.bar(df_test, x = 'Q6', y = 'Percentage', title= 'Education Level for SWEs, Data Scientists, and Data Engineers in 2018', 
             color = 'Q4' , barmode='group', labels={'Percentage':'Percentage (%)'}, opacity = OPACITY) #facet_row='Q1')
fig.show()

Key Takeaways: 
    1. A Master's degree is the most common level of education for  the three selected roles. 43.7% of Software Engineers (SWE), 55.4% of Data Scientists, and 55.76% of Data Engineers have a Master's degree. 
    2. 39.49% of Software Engineers (SWE), 20.18% of Data Scientists, and 29.58% of Data Engineers have a Bachelor's degree. A Master's or a Bachelor's degree is more popular than a Doctorate degree. 
    3. Data engineers have the highest percentage of Master's and Bachelor degree graduates. Data Scientists have the highest percentage of doctorage degrees at 18.81%. 

In [91]:
# Salary Breakdown Per Role Per Level of Education
df_test = df_multi_tech.groupby(['Q6', 'Q9', 'Q4']).size().reset_index(name = 'Percentage')
a = df_test.groupby('Q9')['Percentage'].transform('sum')
df_test['Percentage'] = df_test['Percentage'].div(a).multiply(100)
df_test = df_test.sort_values(by = ['Percentage', 'Q6'], ascending = False)

In [92]:
fig = px.bar(df_test, x = 'Q9', y = 'Percentage', title= 'International Compensation Breakdown by Role and Level of Education in 2018', 
             color = 'Q6' , barmode='group', labels={'Percentage':'Percentage (%)'}, facet_row='Q4', height = 3000,
            color_discrete_sequence=[COLUMN_COLORS[1], COLUMN_COLORS[0], COLUMN_COLORS[2]], opacity = OPACITY)
fig.show()

Notes:
    1. The most common amount international compensation for data scientists, SWEs, and Data Engineers with a Master's degree is between 40-50,000, 20-30,000, and 300-400,000, respectively. 

## How does compensation per role and per level of education vary in the Top-3 Most Represented Countries? 

In [93]:
df_test = df_multi_tech.groupby(['Q6', 'Q9', 'Q4', 'Q3']).size().reset_index(name = 'Percentage')
a = df_test.groupby('Q9')['Percentage'].transform('sum')
df_test['Percentage'] = df_test['Percentage'].div(a).multiply(100)

n = 3
df_placeholder = pd.DataFrame({'count' : df_multi_choice.groupby([ "Q3"] ).size()}).reset_index().sort_values(by = 'count', ascending = False)

top_n_countries = df_placeholder['Q3'].unique()[:n]
top_n_countries

df_test = df_test[df_test['Q3'].isin(top_n_countries)]
df_test = df_test.sort_values(by = ['Percentage'], ascending = False)

In [94]:
fig = px.bar(df_test, x = 'Q9', y = 'Percentage', title= ('Top {} Countries: Compensation Breakdown by Role and Level of Education in 2018').format(n), 
             color = 'Q6' , barmode='group', labels={'Percentage':'Percentage (%)', 'Q9':''}, facet_row='Q4', facet_col = 'Q3', height = 3000,
            color_discrete_sequence=[COLUMN_COLORS[1], COLUMN_COLORS[0], COLUMN_COLORS[2]], opacity = OPACITY)
fig.show()

Notes:
Disclaimer: These results do not segment by the years of experience of a respondent. The percentages are grouped by compensation. 

    1. In the United States, the most common compensation for data scientists with a Master's, Doctoral, or Bachelor's degree is 125-150,000 (28.42%), 200-25000+ (21.79%), and 250-300,000 (12.9%), respectively. In India, the most common compensation for data scientists with a Master's, Doctoral, or Bachelor's degree is 400-500,000 (9.09%), 500,000+ (4.17%), and 500,000+ (11.76%), respectively. In China, the most common compensation for data scientists with a Master's, Doctoral, or Bachelor's degree is 500,000+ (5.88%), 500,000+ (5.88%), and 90-100,000 (0.35%), respectively. Key takeaway: Data Scientist wages in the USA are much different in magnitude than in China and India. As a Data Scientist with advanced education -- with a Master's or Doctoral degree -- in China and India, you are most likely to earn 400,000+. A larger percentage of Data Scientists in the USA earn more than 125,000 than in China and India. Although India and China's most common level of compensation for advanced degree holders is greater than in the USA, there are more jobs in the USA for Data Scientists to earn 125,000+. 
    2. Software Engineer Compensation in Top-3 Countries with Varying Levels of Education: In the United States, the most common compensation for data scientists with a Master's, Doctoral, or Bachelor's degree is 300-400,000 (16.67%), 400-500,000+ (18.18%), and 250-300,000+ (12.90%), respectively. In India, the most common compensation for data scientists with a Master's, Doctoral, or Bachelor's degree is 0-10,000 (6.14%), 20-30,000+ (0.34%), and 0-10,000 (14.36%), respectively. In China, the most common compensation for data scientists with a Master's, Doctoral, or Bachelor's degree is 30-40,000 (2.05%), 40-50,000+ (0.22%), and 30-40,000 (1.23%), respectively. Key takeaway: Software engineering wages in the USA far outweigh China and India. Regardless of education level, the most common compensation level in India and China is within a range of 20,000 from a base of 10,000. In the USA, SWEs with doctoral or Master's level education is more likely to earn greater than 100,000 than SWEs with a bachelor's degree.
    3. Data Engineer Compensation in Top-3 Countries with Varying Levels of Education: In the United States, the most common compensation for data engineers with a Master's, Doctoral, or Bachelor's degree is 125-150,000 (4.79%), 200-250,000 (1.28%), and 150-200,000+ (3.48%), respectively. In India, the most common compensation for data engineers with a Master's, Doctoral, or Bachelor's degree is 0-10,000 (1.29%), 40-50,000 (0.22%), and 10-20,000 (1.57%), respectively. In China, the most common compensation for data engineers with a Master's, Doctoral, or Bachelor's degree is 0-10,000 (1.21%), Not disclosed (0.30%), and 10-20,000 (1.29%), respectively. 

In [95]:
select_audience_segment = ['Software Engineer', 'Data Scientist', 'Data Engineer']


In [96]:
df_edu = df_multi_choice[['Q6', 'Q5', 'Q9']]
df_test = df_edu[(df_edu['Q6'].isin(select_audience_segment))]
df_test = df_test.groupby(['Q6','Q5']).size().reset_index(name = 'Percentage')
a = df_test.groupby('Q6')['Percentage'].transform('sum')
df_test['Percentage'] = df_test['Percentage'].div(a).multiply(100)


df_test = df_test.sort_values(by = ['Percentage'], ascending = False)
df_test.head(2)

fig = px.bar(df_test, x = 'Q5', y = 'Percentage', title= (' Undergraduate Degree Per Role'),
              color = 'Q6', barmode = 'group', labels={'Percentage':'Percentage (%)', 'Q5':''},
            opacity = OPACITY, color_discrete_sequence=[COLUMN_COLORS[0], COLUMN_COLORS[2], COLUMN_COLORS[1]])
fig.show()

Note: 
    1. Apart from computer science degrees, the undergraduate degree most common for data scientists are Mathematics or Statistics at 20.29%, or Engineering (non-computer focused) at 15.67%. Physics is a close 3rd for data scientists at 8.23%. 
    2. For Data Engineers, an undergraduate degree in Engineering (non-computer focused) represents a 10% difference in number of roles reported as compared to data engineers with math or statistics degrees. 
    3. Software engineers are not as often majoring in math or statistics, engineering (non-computer focused) when compared to data engineers and data scientists. 

## HOW ARE DATA SCIENTISTS BECOMING DATA SCIENTISTS? 
2018-(Q6, Q8, Q18, Q24, Q26, Q35, Q36, Q37, Q38, Q39, Q40)


### Review Questino 35, 36, 38, 39 due to multi-part answers

df_multi_choice['Q36_Part_6']

In [97]:
select_audience_segment = ['Software Engineer', 'Data Scientist', 'Data Engineer']
select_geography = top_n_countries

### What are the top programming languages to learn first?

In [98]:
df_data_become = df_multi_choice[['Q6', 'Q3', 'Q8', 'Q9', 'Q17','Q18', 'Q24', 'Q26', 'Q35_Part_1', 'Q36_Part_1', 'Q37', 'Q38_Part_1', 'Q39_Part_1', 'Q40']]# 'Q35'], 'Q36', 'Q37', 'Q38', 'Q39', 'Q40']]

df_test = df_data_become[df_data_become['Q6'].isin(select_audience_segment)]
df_test.head(2)

Q6                        Q3   Q8  \
3  Data Scientist  United States of America  0-1   
6  Data Scientist                  Colombia  0-1   

                                                  Q9                    Q17  \
3                                           0-10,000                   Java   
6  I do not wish to disclose my approximate yearl...  Javascript/Typescript   

      Q18         Q24             Q26 Q35_Part_1 Q36_Part_1           Q37  \
3  Python  5-10 years  Definitely yes          0        NaN      DataCamp   
6  Python   3-5 years           Maybe         40        NaN  Kaggle Learn   

  Q38_Part_1                 Q39_Part_1  \
3    Twitter             Slightly worse   
6        NaN  No opinion; I do not know   

                                                 Q40  
3  Independent projects are equally important as ...  
6  Independent projects are slightly less importa...

In [99]:
df_test = df_test.groupby(['Q18', 'Q6','Q9']).size().reset_index(name = 'Percentage')
a = df_test.groupby('Q6')['Percentage'].transform('sum')
df_test['Percentage'] = df_test['Percentage'].div(a).multiply(100)
df_test = df_test.sort_values(by = ['Percentage'], ascending = False)



In [100]:
fig = px.bar(df_test, x = 'Q18', y = 'Percentage', title= 'Recommended Programming Language to Learn First in 2018', color = 'Q6', 
             barmode='group', labels={'Percentage':'Percentage (%)', 'Q18': 'Programming Language'},
            color_discrete_sequence=[COLUMN_COLORS[0], COLUMN_COLORS[2], COLUMN_COLORS[1]], opacity = OPACITY)
fig.show()

fig = px.bar(df_test, x = 'Q9', y = 'Percentage', title= ('Top {} Countries: Compensation Breakdown by Role and Level of Education in 2018').format(n), 
             color = 'Q6' , barmode='group', labels={'Percentage':'Percentage (%)', 'Q9':''}, facet_row='Q4', facet_col = 'Q3', height = 3000)
fig.show()

In [101]:
df_data_become = df_multi_choice[['Q17', 'Q6']]
df_test = df_data_become[df_data_become['Q6'].isin(select_audience_segment)]

df_test = df_test.groupby(['Q6', 'Q17']).size().reset_index(name = 'Percentage')
a = df_test.groupby('Q6')['Percentage'].transform('sum')
df_test['Percentage'] = df_test['Percentage'].div(a).multiply(100)
df_test = df_test.sort_values(by = ['Percentage'], ascending = False)
df_test.head()
fig = px.bar(df_test, x = 'Q17', y = 'Percentage', title= 'Most used Programming Language in 2018', color = 'Q6', 
             barmode='group', labels={'Percentage':'Percentage (%)', 'Q17': 'Programming Language'},
            color_discrete_sequence=[COLUMN_COLORS[1], COLUMN_COLORS[2], COLUMN_COLORS[0]], opacity = OPACITY)
fig.show()

In [102]:
df_data_become = df_multi_choice[['Q6', 'Q3', 'Q8', 'Q9', 'Q18', 'Q24', 'Q26', 'Q35_Part_1', 'Q36_Part_1', 'Q37', 'Q38_Part_1', 'Q39_Part_1', 'Q40']]# 'Q35'], 'Q36', 'Q37', 'Q38', 'Q39', 'Q40']]
df_test = df_data_become[df_data_become['Q6'].isin(select_audience_segment)]
df_test.head(2)

Q6                        Q3   Q8  \
3  Data Scientist  United States of America  0-1   
6  Data Scientist                  Colombia  0-1   

                                                  Q9     Q18         Q24  \
3                                           0-10,000  Python  5-10 years   
6  I do not wish to disclose my approximate yearl...  Python   3-5 years   

              Q26 Q35_Part_1 Q36_Part_1           Q37 Q38_Part_1  \
3  Definitely yes          0        NaN      DataCamp    Twitter   
6           Maybe         40        NaN  Kaggle Learn        NaN   

                  Q39_Part_1  \
3             Slightly worse   
6  No opinion; I do not know   

                                                 Q40  
3  Independent projects are equally important as ...  
6  Independent projects are slightly less importa...

In [103]:
df_test = df_test.groupby(['Q6', 'Q37']).size().reset_index(name = 'Percentage')
a = df_test.groupby('Q6')['Percentage'].transform('sum')
df_test['Percentage'] = df_test['Percentage'].div(a).multiply(100)
df_test = df_test.sort_values(by = ['Percentage'], ascending = False)

### Which online platform have you spent the most amount of time on?

In [104]:
fig = px.bar(df_test, x = 'Q37', y = 'Percentage', title= ('Top Online Platforms for Time Spent in 2018'), 
             color = 'Q6' , barmode='group', labels={'Percentage':'Percentage (%)', 'Q37':''}, height = 500,
            color_discrete_sequence=[COLUMN_COLORS[1], COLUMN_COLORS[2], COLUMN_COLORS[0]], opacity = OPACITY)
fig.show()

Note: 
 1. For Data Scientists, the top-3 platforms with the most respondents reporting that they spend the most time on are Coursera (45.54%), DataCamp (12.39%), and Udemy (10.2%).
 2. For Data Engineers, the top-3 platforms with the most respondents reporting that they spend the most time on are Coursera (42.82%), Udemy (14.08%), and Udacity (9.19%).
 3. For Software Engineers, the top-3 platforms with the most respondents reporting that they spend the most time on are Coursera (39.52%), Udemy (12.29%), and Udacity (12.29%).

## Which platform with the most time spent has respondents with highest reported salaries (i.e., most candidates earning 6-figure incomes)?

In [105]:
select_salaries = ['500,000+', '400-500,000', '300-400,000', '250-300,000',
                  '200-250,000', '100-125,000', '150-200,000']

In [106]:
df_data_become = df_multi_choice[['Q6', 'Q3', 'Q8', 'Q9', 'Q18', 'Q24', 'Q26', 'Q35_Part_1', 'Q36_Part_1', 'Q37', 'Q38_Part_1', 'Q39_Part_1', 'Q40']]# 'Q35'], 'Q36', 'Q37', 'Q38', 'Q39', 'Q40']]
df_test = df_data_become[(df_data_become['Q6'].isin(select_audience_segment)) &
                        df_data_become['Q9'].isin(select_salaries)]
df_test.head(2)

Q6                        Q3     Q8           Q9     Q18  \
33     Data Scientist               Netherlands    0-1  100-125,000  Python   
41  Software Engineer  United States of America  10-15  400-500,000  Python   

                                              Q24             Q26 Q35_Part_1  \
33                                      3-5 years  Definitely yes          0   
41  I have never written code but I want to learn  Definitely not        NaN   

   Q36_Part_1       Q37 Q38_Part_1  Q39_Part_1  \
33    Udacity  Coursera        NaN  Much worse   
41        NaN       NaN        NaN  Much worse   

                                                  Q40  
33  Independent projects are slightly less importa...  
41  Independent projects are slightly less importa...

In [107]:
df_test = df_test.groupby(['Q6', 'Q9', 'Q37']).size().reset_index(name = 'Percentage')
a = df_test.groupby('Q6')['Percentage'].transform('sum')
df_test['Percentage'] = df_test['Percentage'].div(a).multiply(100)
df_test = df_test.sort_values(by = ['Percentage'], ascending = False)

In [108]:
fig = px.bar(df_test, x = 'Q37', y = 'Percentage', title= ('Top in 2018'), 
             color = 'Q6' , barmode='group', labels={'Percentage':'Percentage (%)', 'Q37':''}, facet_row = 'Q9', height = 1000,
            color_discrete_sequence=[COLUMN_COLORS[2], COLUMN_COLORS[1], COLUMN_COLORS[0]], opacity = OPACITY)
fig.show()

Note:
    1. In aggregate, Coursera is the best producing platform in terms of the likelihood of having a six-figure salarY if it's the platform you spend the most time on. Disclaimer: this doesn't translate into the best platform for the number of dollars generated per total time spent. For example, 28.9%, 27.88%, and 23% of Data Engineers, Data Scientists, and Software Engineers, respectively, reported having at least a 100,000 salary after reporting that Coursera was the online platform that they spend the most time on. 

fig = px.bar(df_test, x = 'Q37', y = 'Percentage', title= ('Top in 2018'), 
             color = 'Q6' , barmode='group', labels={'Percentage':'Percentage (%)', 'Q37':''}, facet_col = 'Q3', height = 500)
fig.show()

## How does number of years of experience writing code to analyze data relate to compensation?

In [109]:
df_data_become = df_multi_choice[['Q6', 'Q3', 'Q8', 'Q9', 'Q18', 'Q24', 'Q26', 'Q35_Part_1', 'Q36_Part_1', 'Q37', 'Q38_Part_1', 'Q39_Part_1', 'Q40']]# 'Q35'], 'Q36', 'Q37', 'Q38', 'Q39', 'Q40']]
df_test = df_data_become[(df_data_become['Q6'].isin(select_audience_segment))]
df_test.head(2)

Q6                        Q3   Q8  \
3  Data Scientist  United States of America  0-1   
6  Data Scientist                  Colombia  0-1   

                                                  Q9     Q18         Q24  \
3                                           0-10,000  Python  5-10 years   
6  I do not wish to disclose my approximate yearl...  Python   3-5 years   

              Q26 Q35_Part_1 Q36_Part_1           Q37 Q38_Part_1  \
3  Definitely yes          0        NaN      DataCamp    Twitter   
6           Maybe         40        NaN  Kaggle Learn        NaN   

                  Q39_Part_1  \
3             Slightly worse   
6  No opinion; I do not know   

                                                 Q40  
3  Independent projects are equally important as ...  
6  Independent projects are slightly less importa...

In [110]:
df_test = df_test.groupby(['Q6', 'Q24']).size().reset_index(name = 'Percentage')
a = df_test.groupby('Q6')['Percentage'].transform('sum')
df_test['Percentage'] = df_test['Percentage'].div(a).multiply(100)
df_test = df_test.sort_values(by = ['Percentage'], ascending = False)

In [111]:
fig = px.bar(df_test, x = 'Q24', y = 'Percentage', title= ('Number of Years Writing Code to Analyze Data 2018'), 
             color = 'Q6' , barmode='group', labels={'Percentage':'Percentage (%)', 'Q38_Part_1':''},
            color_discrete_sequence=[COLUMN_COLORS[0], COLUMN_COLORS[2], COLUMN_COLORS[1]], opacity = OPACITY)#, facet_row = 'Q9', height = 1000)
fig.show()

Note: 
 1. A majority of our respondents for data scientists, data engineers, and software engineers have less than or equal to 10 years of experience writing code to analyze data. 

In [112]:
select_years_segment = ['< 1 year', '1-2 years', '3-5 years', '5-10 years']

In [113]:
df_data_become = df_multi_choice[['Q6', 'Q3', 'Q8', 'Q9', 'Q18', 'Q24', 'Q26', 'Q35_Part_1', 'Q36_Part_1', 'Q37', 'Q38_Part_1', 'Q39_Part_1', 'Q40']]# 'Q35'], 'Q36', 'Q37', 'Q38', 'Q39', 'Q40']]
df_test = df_data_become[(df_data_become['Q6'].isin(select_audience_segment)) &
                         (df_data_become['Q24'].isin(select_years_segment)) &
                         (df_data_become['Q9'].isin(select_salaries) & 
                         (df_data_become['Q3'].isin(select_geography)))
                          ]
df_test.head(2)

df_test = df_test.groupby(['Q6', 'Q24', 'Q9', 'Q3']).size().reset_index(name = 'Percentage')
a = df_test.groupby('Q6')['Percentage'].transform('sum')
df_test['Percentage'] = df_test['Percentage'].div(a).multiply(100)
df_test = df_test.sort_values(by = ['Percentage'], ascending = False)

In [114]:
fig = px.bar(df_test, x = 'Q24', y = 'Percentage', title= ('Salaries for Data Roles Across Top-3 Respondent Countries versus Years of Experience in 2018'), 
             color = 'Q6' , barmode='group', labels={'Percentage':'Percentage (%)', 'Q38_Part_1':''}, facet_row = 'Q9', facet_col = 'Q3', height = 2000,
            color_discrete_sequence=[COLUMN_COLORS[2], COLUMN_COLORS[1], COLUMN_COLORS[0]], opacity = OPACITY)
fig.show()

Note: 
    1. 

## Best way to demonstrate expertise in data science?

In [115]:
df_data_become = df_multi_choice[['Q6', 'Q3', 'Q8', 'Q9', 'Q18', 'Q24', 'Q26', 'Q35_Part_1', 'Q36_Part_1', 'Q37', 'Q38_Part_1', 'Q39_Part_1', 'Q40']]# 'Q35'], 'Q36', 'Q37', 'Q38', 'Q39', 'Q40']]
df_test = df_data_become[(df_data_become['Q6'].isin(select_audience_segment))] #&
df_test = df_test.groupby(['Q6', 'Q40']).size().reset_index(name = 'Percentage')
a = df_test.groupby('Q6')['Percentage'].transform('sum')
df_test['Percentage'] = df_test['Percentage'].div(a).multiply(100)
df_test = df_test.sort_values(by = ['Percentage'], ascending = False)

fig = px.bar(df_test, x = 'Q40', y = 'Percentage', title= (' 2018'), 
             color = 'Q6' , barmode='group', labels={'Percentage':'Percentage (%)', 'Q40':''},
            color_discrete_sequence=[COLUMN_COLORS[2], COLUMN_COLORS[0], COLUMN_COLORS[1]], opacity = OPACITY)#, facet_row = 'Q9', facet_col = 'Q3', height = 2000)
fig.show()

Note:
    1. It's clear that projects are the best way to break into the industry is to do projects. 

## How do you perceive the quality of MOOCs and in-person bootcamps to quality of education from traditional brick & mortar institutions?

In [116]:
df_data_become = df_multi_choice[['Q6', 'Q3', 'Q8', 'Q9', 'Q18', 'Q24', 'Q26', 'Q35_Part_1', 'Q36_Part_1', 'Q37', 'Q38_Part_1', 'Q39_Part_1', 'Q40']]# 'Q35'], 'Q36', 'Q37', 'Q38', 'Q39', 'Q40']]
df_test = df_data_become[(df_data_become['Q6'].isin(select_audience_segment))&
                         (df_data_become['Q3'].isin(select_geography))&
                        (df_data_become['Q9'].isin(select_salaries))] 
    
df_test = df_test.groupby(['Q6', 'Q3', 'Q39_Part_1', 'Q9', 'Q37']).size().reset_index(name = 'Percentage')
a = df_test.groupby('Q6')['Percentage'].transform('sum')
df_test['Percentage'] = df_test['Percentage'].div(a).multiply(100)
df_test = df_test.sort_values(by = ['Q39_Part_1', 'Percentage'], ascending = False)

fig = px.bar(df_test, x = 'Q39_Part_1', y = 'Percentage', title= ('Perception of MOOCs and Bootcamps to Traditional Brick & Mortar Education in 2018'), 
             color = 'Q6' , barmode='group', labels={'Percentage':'Percentage (%)', 'Q39_Part_1':''},
            color_discrete_sequence=[COLUMN_COLORS[2], COLUMN_COLORS[1], COLUMN_COLORS[0]], opacity = OPACITY)#, facet_row = 'Q9', facet_col = 'Q3', height = 2000)
fig.show()

Note:

In 2018
    1. Data Scientists: ~9.7% of Data Scientists rate MOOCs are Much Worse than traditional education. Previous data also shows. 15% report it to be Much Better. 
    2. Data Engineers are the only group that did not report any MOOCs as Much Worse. 43% report them as slightly better, 22% report it as slightly worse, and 14% report MOOCs as Much Better. This leads us to believe that MOOCs for data engineers is generally better than Brick & Mortar education.

#### The International Perspective Top 3 Survey Respondent Countries that are Data Scientists, Data Engineers, or Software Engineers

In [117]:
df_data_become = df_multi_choice[['Q6', 'Q3', 'Q8', 'Q9', 'Q18', 'Q24', 'Q26', 'Q35_Part_1', 'Q36_Part_1', 'Q37', 'Q38_Part_1', 'Q39_Part_1', 'Q40']]# 'Q35'], 'Q36', 'Q37', 'Q38', 'Q39', 'Q40']]
df_test = df_data_become[(df_data_become['Q6'].isin(select_audience_segment))&
                         (df_data_become['Q3'].isin(select_geography))&
                        (df_data_become['Q9'].isin(select_salaries))] 

df_test = df_test.groupby(['Q6','Q3', 'Q39_Part_1', 'Q9', 'Q37']).size().reset_index(name = 'Percentage')
a = df_test.groupby('Q6')['Percentage'].transform('sum')
df_test['Percentage'] = df_test['Percentage'].div(a).multiply(100)
df_test = df_test.sort_values(by = ['Q39_Part_1', 'Percentage'], ascending = False)

fig = px.bar(df_test, x = 'Q39_Part_1', y = 'Percentage', title= ('Perception of MOOCs and Bootcamps to Traditional Brick & Mortar Education in 2018'), 
              barmode='group', labels={'Percentage':'Percentage (%)', 'Q39_Part_1':''}, facet_col = 'Q3', 
             height = 500, opacity = OPACITY)
fig.show()

Note:
    1. In India and in China, the data tends to skew towards agreeing that MOOCs are better than traditional brick and mortar institutions. 
    2. In the United States, the answer tends to fall closer to a normal distribution. There is however a larger percent of the population for our segmented technical audience that believes MOOCs are Much Better as opposed to Much Worse when compared to traditional education. 

In [118]:
df_test = df_test.sort_values(by = ['Q39_Part_1', 'Percentage'], ascending = False)
fig = px.bar(df_test, x = 'Q39_Part_1', y = 'Percentage', title= (' 2018'), 
              barmode='group', labels={'Percentage':'Percentage (%)', 'Q39_Part_1':''}, facet_col = 'Q37', 
             height = 500, width = 2000, opacity = OPACITY)
fig.show()

Note:
    We wanted to understand how our segment audience ranks different online platforms to traditional brick and mortar institutions. 
    1. Note: 26% of our segment audience that spends the most time on Coursera believes that the content is Much Better than traditional Brick & Mortar education. 10% of Respondents believe that Udemy is Much Better. 4% of Udacity students believe it is Much Better. 4% of DataCamp students believe it is much better.  

In [119]:
select_MOOCs = ['Coursera', 'Online University Courses', 'Udacity', 'DataCamp', 'Udemy']

In [120]:
df_data_become = df_multi_choice[['Q6', 'Q3', 'Q8', 'Q9', 'Q18', 'Q24', 'Q26', 'Q35_Part_1', 'Q36_Part_1', 'Q37', 'Q38_Part_1', 'Q39_Part_1', 'Q40']]# 'Q35'], 'Q36', 'Q37', 'Q38', 'Q39', 'Q40']]
df_test = df_data_become[(df_data_become['Q6'].isin(select_audience_segment))&
                         (df_data_become['Q3'].isin(select_geography))&
                        (df_data_become['Q9'].isin(select_salaries)) &
                        (df_data_become['Q37'].isin(select_MOOCs))] 

df_test = df_test.groupby(['Q6','Q3', 'Q39_Part_1', 'Q9', 'Q37']).size().reset_index(name = 'Percentage')
a = df_test.groupby('Q6')['Percentage'].transform('sum')
df_test['Percentage'] = df_test['Percentage'].div(a).multiply(100)

df_test = df_test.sort_values(by = ['Q39_Part_1', 'Percentage', 'Q9'], ascending = False)
fig = px.bar(df_test, x = 'Q39_Part_1', y = 'Percentage', title= (' 2018'),
             barmode='group', labels={'Percentage':'Percentage (%)', 'Q39_Part_1':''}, facet_row = 'Q9',
             facet_col = 'Q37', height = 1200, width = 1000, opacity = OPACITY)
fig.show()

## WHAT TOOLS/PROGRAMMING LANGUAGES/ENVIRONMENTS DO DATA SCIENTISTS USE AND HOW HAS THAT CHANGED OVER TIME? 

2018-(Q12, Q13, Q14, Q15, Q16, Q17, Q19, Q20, Q21, Q22, Q49, Q27, Q28, Q29, Q30, Q31, Q32, Q33, Q49, Q50)
6.) Most used Programming Language? What languages are recommended to learn?
7.) Is data science happening more often in the cloud or locally? What type of IDEs?
3.) Tools they're using

## Q12, Q13, Q14, Q15, Q16, Q17, Q19, Q21, Q22, Q27, Q28, Q30, Q31, Q33, Q49, Q50

df_data_tools = df_multi_choice[['Q6', 'Q3', 'Q8', 'Q9', 'Q12', 'Q13', 
                                 'Q14', 'Q15', 'Q16', 'Q17', 'Q19', 'Q20',
                                 'Q21','Q22','Q27','Q28','Q29','Q30','Q31','Q32', 'Q33','Q49','Q50']]# 'Q35'], 'Q36', 'Q37', 'Q38', 'Q39', 'Q40']]
df_test = df_data_tools[(df_data_tools['Q6'].isin(select_audience_segment))]
                         #(df_data_tools['Q3'].isin(select_geography))&
                        #(df_data_tools['Q9'].isin(select_salaries)) &
                        #(df_data_tools['Q37'].isin(select_MOOCs))] 

## Bias in Data:

BIAS IN DATA: WHAT ARE WE DOING TO ADDRESS IT? 2018-(Q41, Q43, Q45, Q46, Q48)
1.) Do DS/ML believe in data bias in their algorithms?

In [121]:
## q41, Q45 are multipart entries

In [122]:
df_bias = df_multi_choice[['Q6', 'Q3', 'Q43', 'Q46', 'Q48']]
df_test = df_bias[(df_bias['Q6'].isin(select_audience_segment))]
df_test = df_test.groupby(['Q6', 'Q43']).size().reset_index(name = 'Percentage')
a = df_test.groupby('Q6')['Percentage'].transform('sum')
df_test['Percentage'] = df_test['Percentage'].div(a).multiply(100)


df_test = df_test.sort_values(by = ['Q6'], ascending = False)
df_test.head(2)

Q6     Q43  Percentage
32  Software Engineer  90-100    1.748907
31  Software Engineer   80-90    0.999375

In [123]:
fig = px.bar(df_test, x = 'Q43', y = 'Percentage', title= (' Percent of Time Spent Exploring Unfair Bias in Data or Model'),
              color = 'Q6', barmode = 'group', labels={'Percentage':'Percentage (%) of Respondents', 'Q43':'Percent Time Spent Exploring Bias'},
              color_discrete_sequence=COLUMN_COLORS,
              opacity = OPACITY)
fig.show()

In [124]:
df_bias = df_multi_choice[['Q6', 'Q3', 'Q43', 'Q46', 'Q48']]
df_test = df_bias[(df_bias['Q6'].isin(select_audience_segment))]
df_test = df_test.groupby(['Q6', 'Q46']).size().reset_index(name = 'Percentage')
a = df_test.groupby('Q6')['Percentage'].transform('sum')
df_test['Percentage'] = df_test['Percentage'].div(a).multiply(100)


df_test = df_test.sort_values(by = ['Q6'], ascending = False)
df_test.head(2)

fig = px.bar(df_test, x = 'Q46', y = 'Percentage', title= (' Percent of Time Spent Exploring Model Insights Per Role in 2018'),
              color = 'Q6', barmode = 'group', labels={'Percentage':'Percentage (%) of Total Respondents', 'Q46': 'Percent Time Exploring Model Insights'},
            opacity = OPACITY,
            color_discrete_sequence=COLUMN_COLORS)
fig.show()

Note: 
    1. A larger proportion of data scientists spend more than 50% of their time exploring model insights as compared to data engineers and software engineers. Software engineers spend the least amount of time reviewing model insights. A larger proportion of each group, however, spends less than 20-30% of their time reviewing model insights.

In [125]:
df_bias = df_multi_choice[['Q6', 'Q3', 'Q43', 'Q46', 'Q48']]
df_test = df_bias[(df_bias['Q6'].isin(select_audience_segment))]
df_test = df_test.groupby(['Q6', 'Q48']).size().reset_index(name = 'Percentage')
a = df_test.groupby('Q6')['Percentage'].transform('sum')
df_test['Percentage'] = df_test['Percentage'].div(a).multiply(100)


df_test = df_test.sort_values(by = ['Percentage'], ascending = False)
df_test.head(2)

fig = px.bar(df_test, x = 'Q48', y = 'Percentage', title= (' Are Machine Learning Models "Black Boxes"'),
              color = 'Q6', barmode = 'group', labels={'Percentage':'Percentage (%) of Total Respondents', 'Q46': 'Percent Time Exploring Model Insights'},
            opacity = OPACITY, height = 800, width = 800, color_discrete_sequence=[COLUMN_COLORS[1], COLUMN_COLORS[2], COLUMN_COLORS[0]] )
fig.show()

## INDUSTRY BREAKDOWN: WHERE ARE THE MOST DATA SCIENTISTS? IS IT INDUSTRY AGNOSTIC?
2018-(Q7, Q10, Q42, Q44)

In [126]:
#Q42 and Q44 are multiple part

In [127]:
df_industry = df_multi_choice[['Q3', 'Q10']]#, 'Q42', 'Q44']]
df_test = df_industry[(df_industry['Q3'].isin(select_geography))]
df_test = df_test.groupby(['Q3', 'Q10']).size().reset_index(name = 'Percentage')
a = df_test.groupby('Q3')['Percentage'].transform('sum')
df_test['Percentage'] = df_test['Percentage'].div(a).multiply(100)
df_test

Q3  \
0                      China   
1                      China   
2                      China   
3                      China   
4                      China   
5                      China   
6                      India   
7                      India   
8                      India   
9                      India   
10                     India   
11                     India   
12  United States of America   
13  United States of America   
14  United States of America   
15  United States of America   
16  United States of America   
17  United States of America   

                                                  Q10  Percentage  
0                                       I do not know   18.168280  
1                       No (we do not use ML methods)   13.402829  
2   We are exploring ML methods (and may one day p...   28.518243  
3   We have well established ML methods (i.e., mod...    9.009680  
4   We recently started using ML methods (i.e., mo...   18.912882  
5   We use ML methods for generating insights (but...   11.988086  
6                                       I do not know   21.489589  
7                       No (we do not use ML methods)   20.315003  
8   We are exploring ML methods (and may one day p...   23.571810  
9   We have well established ML methods (i.e., mod...    8.969568  
10  We recently started using ML methods (i.e., mo...   17.618793  
11  We use ML methods for generating insights (but...    8.035238  
12                                      I do not know   12.655205  
13                      No (we do not use ML methods)   18.935053  
14  We are exploring ML methods (and may one day p...   19.054441  
15  We have well established ML methods (i.e., mod...   19.985673  
16  We recently started using ML methods (i.e., mo...   17.669532  
17  We use ML methods for generating insights (but...   11.700096

In [128]:
COUNTRY_COLOR = ['RED', 'ORANGE', 'BLUE']
df_industry = df_multi_choice[['Q3', 'Q10']]#, 'Q42', 'Q44']]
df_test = df_industry[(df_industry['Q3'].isin(select_geography))]
df_test = df_test.groupby(['Q3', 'Q10']).size().reset_index(name = 'Percentage')
a = df_test.groupby('Q3')['Percentage'].transform('sum')
df_test['Percentage'] = df_test['Percentage'].div(a).multiply(100)

fig = px.bar(df_test, x = 'Q10', y = 'Percentage', title= (' Does Current Employer Use ML in their Business (2018)?"'),
            color = 'Q3', barmode = 'group', labels={'Percentage':'Percentage (%)', 'Q10': ''},
            opacity = OPACITY, height = 500, width = 1000,  color_discrete_sequence=COUNTRY_COLOR)
fig.show()